src/Altinn.App.Api/Controllers/DataController.cs

```diff
@@ -353,6 +366,167 @@ private static bool FileValidationEnabledForDataType(DataType dataTypeFromMetada
+
+                if (dataElement == null)
+                {
+                    return NotFound("Did not find data element");
+                }
+
+                var dataType = await GetDataType(dataElement);
+
+                if (dataType?.AppLogic?.ClassRef is null)
+                {
+                    _logger.LogError(
+                        "Could not determine if {dataType} requires app logic for application {org}/{app}",
+                        dataType,
+                        org,
+                        app);
+                    return BadRequest($"Could not determine if data type {dataType?.Id} requires application logic.");
+                }
+
+                var modelType = _appModel.GetModelType(dataType.AppLogic.ClassRef);
+
+                var oldModel =
+                    await _dataClient.GetFormData(instanceGuid, modelType, org, app, instanceOwnerPartyId, dataGuid);
+
+                var (response, problemDetails) =
+                    await PatchFormDataImplementation(dataType, dataElement, dataPatchRequest, oldModel, instance);
+
+                if (problemDetails is not null)
+                {
+                    return StatusCode(problemDetails.Status ?? 500, problemDetails);
+                }
+
+                await UpdatePresentationTextsOnInstance(instance, dataType.Id, response.NewDataModel);
+                await UpdateDataValuesOnInstance(instance, dataType.Id, response.NewDataModel);
+
+                // Save Formdata to database
+                await _dataClient.UpdateData(
+                    response.NewDataModel,
+                    instanceGuid,
+                    modelType,
+                    org,
+                    app,
+                    instanceOwnerPartyId,
+                    dataGuid);
+
+                return Ok(response);
+            }
+            catch (PlatformHttpException e)
+            {
+                return HandlePlatformHttpException(e, $"Unable to update data element {dataGuid} for instance {instanceOwnerPartyId}/{instanceGuid}");
+            }
+        }
+
+        /// <summary>
+        /// Part of <see cref="PatchFormData" /> that is separated out for testing purposes.
+        /// </summary>
+        /// <param name="dataType">The type of the data element</param>
+        /// <param name="dataElement">The data element</param>
+        /// <param name="dataPatchRequest">Container object for the <see cref="JsonPatch" /> and list of ignored validators</param>
+        /// <param name="oldModel">The old state of the form data</param>
+        /// <param name="instance">The instance</param>
+        /// <returns>DataPatchResponse after this patch operation</returns>
+        internal async Task<(DataPatchResponse, ProblemDetails?)> PatchFormDataImplementation(DataType dataType, DataElement dataElement, DataPatchRequest dataPatchRequest, object oldModel, Instance instance)
```

Comment:
Yes, it could. I made an attempt, but it was hard to write a good summary for what that service class would do. and draw reasonable separation lines.

```c#
/// <summary>
/// Service class for the <see cref="DataController.PatchFormData"> because the controller was getting "rather big"
/// </summary>
```

Doesn't really seem to justify why an extra file is required.

Reply:
I tried, but I don't think the code organisation is improved. 6d51c4db125bd6ec1b35a24b4dcbd572e127f2be

Reply:
I think it makes sense as we don't need the InternalsVisibleFor in the project file. Making the code easier to test is IMO good enough reason to split the code into a separate class